### PR TITLE
Add Python 3.12 support

### PR DESCRIPTION
**Pull request by freakboy3742**
With the release of 3.12.0a1, we can start testing for 3.12 support.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: rubicon/objc/ctypes_patch.py
Comment (freakboy3742) [inline rubicon/objc/ctypes_patch.py:73]: These changes are all Black related; see #230

**Review by dgelessus — APPROVED** [review]
LGTM. Though I think we should wait with this until the Black PR is merged and then rebase, to avoid pointless changes/conflicts in the formatting.

**Review by mhsmith — CHANGES_REQUESTED** [review]
See comments on https://github.com/beeware/briefcase/pull/965.

File: .github/workflows/ci.yml
Comment (mhsmith) [inline .github/workflows/ci.yml:92]: 3.7 is missing here.

Also, there's some inconsistency in the minimum supported version. ci.yml and tox.ini say 3.5, while ctypes_patch.py says 3.4. It may have been tested with 3.4 in the past, but it isn't anymore, so it would be misleading for the message to say that.

I suggest we set all of these to the minimum version supported by CPython itself, which is currently 3.7.

**Review by mhsmith — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml
Comment (dgelessus) [reply .github/workflows/ci.yml:92]: 3.7 is intentionally omitted here, because it's already tested by the other jobs (`smoke` and `macOS-versions`). Might be good to put this in a comment though.

Yeah, it would make sense to update ctypes_patch to the same minimum Python version listed elsewhere. If someone really wants to use the code on 3.4 for some reason, they can ignore the warning or patch it out locally.

#230 already drops Python 3.5 and adjusts the CI config accordingly, so we could wait until that is merged and then rebase/update this PR. (I don't think we would gain much from also dropping 3.6, and AFAICT it's trivial to keep it in the test matrix for now.)

Comment (freakboy3742) [reply .github/workflows/ci.yml:92]: I've set 3.6 as the minimum everywhere; maintaining 3.6 compatibility isn't really costing any extra effort at this point, so we might as well keep it until it becomes onerous to support it (e.g., we dropped 3.5 because it prevented us adopting f-strings). I'll wait until #230 is done before pushing for any further reviews.

**Review by dgelessus — APPROVED** [review]
:+1:

**Review by mhsmith — APPROVED** [review]

**Pull request merged**
